services/wallet/concurrent.go

```diff
@@ -45,17 +45,56 @@ type TransferDownloader interface {
 	GetTransfersByNumber(context.Context, *big.Int) ([]Transfer, error)
 }
 
+var balancesCache = map[common.Address]map[*big.Int]*big.Int{}
+var balanaceMapLock = sync.RWMutex{}
+
+func readCachedBalance(account common.Address, blockNumber *big.Int) (*big.Int, bool) {
+	balanaceMapLock.RLock()
+	defer balanaceMapLock.RUnlock()
+
+	balance, exists := balancesCache[account][blockNumber]
+	return balance, exists
+}
+
+func writeCachedBalance(account common.Address, blockNumber *big.Int, balance *big.Int) {
```

Comment:
```suggestion
func addBalanceToCache(account common.Address, blockNumber *big.Int, balance *big.Int) {
```
because write something a bit different.